### PR TITLE
fix(GameProvider): prevent creating game with duplicate name

### LINKED ISSUE #3982: Game world gets corrupted when creating a new game with the same name
### What you were trying to do

Create a new world by using a name that already belonged to a previously created game on your computer.

### What actually happened

There was no new world created, instead, the first world with that name changed. The modified world was really messed up (see screenshots below).

### How to reproduce

Create two new worlds using the same name. The bug will occur after loading the second world.

### Comments

The game crashed after some time into the corrupted world.

### Screenshots

The world generated after following the "How to reproduce" steps:

#### First test:
Terasology-200523201403-2560x1377

#### Second test:
Terasology-200523201457-2560x1377

#### After creating a 3rd game with the same name:
Terasology-200523204146-2560x1377

### PR DESCRIPTION
**Pull request by DarkWeird**
### Contains
Fixes #3982 

Custom game name works like Default game name prefix ("Game") now.
If you create game with name as previously created new game will have name "<previous> <index>"
If previously created game name already have index, then that index will be incremented.

### How to test

1. create game with name "X"
2. exit to main menu
3. create game with name "X" again
4. check save games. You have save games "X" and "X 1" now.

### CODE REVIEW

File: engine-tests/src/test/java/org/terasology/rendering/nui/layers/mainMenu/savedGames/GameProviderTest.java [hunk 1]
```diff
@@ -162,6 +178,15 @@ public void getNextGameNameDefaultNoSavesTest() {
         assertEquals(GAME_1, name);
     }
 
+    @Test
+    public void getNextGameNameCustomNoSavesTest() {
+        String gameName = "Custom";
+        final String name = GameProvider.getNextGameName(gameName);
+
+        assertNotNull(name);
```
Comment (skaldarnar) [inline engine-tests/src/test/java/org/terasology/rendering/nui/layers/mainMenu/savedGames/GameProviderTest.java:186]: isn't this check included in the next one?

File: engine-tests/src/test/java/org/terasology/rendering/nui/layers/mainMenu/savedGames/GameProviderTest.java [hunk 2]
```diff
@@ -175,6 +200,21 @@ public void getNextGameNameNumberTest() throws IOException {
...
     }
 
+    @ParameterizedTest(name = "getNextGameName(\"{0}\") -> \"{1}\"")
+    @MethodSource("nextGameNamesProvider")
+    public void getNextGameNameNumberCustomNameTest(String gameName, String nextGameName) throws IOException {
+
+        Path customSaveFolder = TMP_SAVES_FOLDER_PATH.resolve(gameName);
+        Files.createDirectories(customSaveFolder);
+        Path manifestFilePath = customSaveFolder.resolve(GameManifest.DEFAULT_FILE_NAME);
+        writeToFile(manifestFilePath, MANIFEST_EXAMPLE.replace(GAME_1, gameName));
```
Comment (skaldarnar) [inline engine-tests/src/test/java/org/terasology/rendering/nui/layers/mainMenu/savedGames/GameProviderTest.java:210]: What do you think of creating a method for this called `mimicSaveGame(customSavesFolder, gameName)`?

File: engine/src/main/java/org/terasology/rendering/nui/layers/mainMenu/savedGames/GameProvider.java [hunk 3]
```diff
@@ -110,21 +111,42 @@ public static boolean isSavesFolderEmpty() {
...
     /**
-     * Generates the game name based on the game number of the last saved game
+     * Generates the game name based on the game number of the last saved game Uses {@link
+     * GameProvider#DEFAULT_GAME_NAME_PREFIX} for resolve.
      */
     public static String getNextGameName() {
+        String nextGameName = getNextGameName(DEFAULT_GAME_NAME_PREFIX);
+        if (nextGameName.equals(DEFAULT_GAME_NAME_PREFIX)) {
+            return nextGameName + 1;
+        }
```
Comment (skaldarnar) [inline engine/src/main/java/org/terasology/rendering/nui/layers/mainMenu/savedGames/GameProvider.java:121]: Why is this necessary? shouldn't this be covered by `getNextGameName(gameName)`?

File: engine/src/main/java/org/terasology/rendering/nui/layers/mainMenu/savedGames/GameProvider.java [hunk 4]
```diff
@@ -110,21 +111,42 @@ public static boolean isSavesFolderEmpty() {
...
+     * @param gameName will to use as game prefix, if saves contains this game name
+     * @return next game name with number. Example: "Game" -> "Game 1", "Game 1 -> Game 2"
+     */
+    public static String getNextGameName(String gameName) {
         int gameNumber = 1;
+        String name = gameName;
         for (GameInfo info : GameProvider.getSavedGames()) {
-            if (info.getManifest().getTitle().startsWith(DEFAULT_GAME_NAME_PREFIX)) {
-                String remainder = info.getManifest().getTitle().substring(DEFAULT_GAME_NAME_PREFIX.length());
+            if (info.getManifest().getTitle().startsWith(name)) {
```
Comment (skaldarnar) [inline engine/src/main/java/org/terasology/rendering/nui/layers/mainMenu/savedGames/GameProvider.java:135]: This prefix match and handling looks a bit shaky to me. We probably should define way more tests to define what we expect here...

File: engine/src/main/java/org/terasology/rendering/nui/layers/mainMenu/savedGames/GameProvider.java (hunk 3 above)
Comment (DarkWeird) [reply engine/src/main/java/org/terasology/rendering/nui/layers/mainMenu/savedGames/GameProvider.java:122]: Default name is "Game 1"
If move this it getNextGameName(gameName) then any custom gamenames will have suffix " 1"

File: engine-tests/src/test/java/org/terasology/rendering/nui/layers/mainMenu/savedGames/GameProviderTest.java (hunk 1 above)
Comment (DarkWeird) [reply engine-tests/src/test/java/org/terasology/rendering/nui/layers/mainMenu/savedGames/GameProviderTest.java:186]: No.we already have "saved" game  in next test . there - not.

**Comment by skaldarnar** [comment]
```
org.opentest4j.AssertionFailedError: expected: <Custom 3> but was: <Custom 2>
	at org.junit.jupiter.api.AssertionUtils.fail(AssertionUtils.java:55)
	at org.junit.jupiter.api.AssertionUtils.failNotEqual(AssertionUtils.java:62)
	at org.junit.jupiter.api.AssertEquals.assertEquals(AssertEquals.java:182)
	at org.junit.jupiter.api.AssertEquals.assertEquals(AssertEquals.java:177)
	at org.junit.jupiter.api.Assertions.assertEquals(Assertions.java:1124)
	at org.terasology.rendering.nui.layers.mainMenu.savedGames.GameProviderTest.getNextGameNameWithNumber(GameProviderTest.java:239)
```

This test succeeds locally 🤔

**Comment by jdrueckert** [comment]
[Game Save Name Override](https://trello.com/c/cfqR4fzV/15-game-save-name-override)

**Pull request merged**